prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,470 @@
+			efa_cntr_report_tx_completion(&base_ep->util_ep, cq_entry.flags);
+			break;
+		case IBV_WC_RECV:
+			efa_cq_handle_rx_completion(base_ep, cq->ibv_cq.ibv_cq_ex, &cq_entry);
+			efa_cntr_report_rx_completion(&base_ep->util_ep, cq_entry.flags);
+			break;
+		case IBV_WC_RECV_RDMA_WITH_IMM:
+			efa_cq_proc_ibv_recv_rdma_with_imm_completion(
+				base_ep, cq->ibv_cq.ibv_cq_ex);
+			efa_cntr_report_rx_completion(&base_ep->util_ep, cq_entry.flags);
+			break;
+		default:
+			EFA_WARN(FI_LOG_EP_CTRL,
+				"Unhandled cq type\n");
+			assert(0 && "Unhandled cq type");
+		}
+
+		num_cqe++;
+		if (num_cqe == cqe_to_process) {
+			break;
+		}
+
+		err = ibv_next_poll(cq->ibv_cq.ibv_cq_ex);
+	}
+
+	if (err && err != ENOENT) {
+		err = err > 0 ? err : -err;
+		prov_errno = ibv_wc_read_vendor_err(cq->ibv_cq.ibv_cq_ex);
+		EFA_WARN(FI_LOG_CQ,
+			 "Unexpected error when polling ibv cq, err: %s (%zd) "
+			 "prov_errno: %s (%d)\n",
+			 fi_strerror(err), err, efa_strerror(prov_errno),
+			 prov_errno);
+		efa_show_help(prov_errno);
+		err_entry = (struct fi_cq_err_entry) {
+			.err = err,
+			.prov_errno = prov_errno,
+			.op_context = NULL,
+		};
+		ofi_cq_write_error(&cq->util_cq, &err_entry);
+	}
+
+	if (should_end_poll)
+		ibv_end_poll(cq->ibv_cq.ibv_cq_ex);
+}
+
+static const char *efa_cq_strerror(struct fid_cq *cq_fid,
+				   int prov_errno,
+				   const void *err_data,
+				   char *buf, size_t len)
+{
+	return err_data
+		? (const char *) err_data
+		: efa_strerror(prov_errno);
+}
+
+static struct fi_ops_cq efa_cq_ops = {
+	.size = sizeof(struct fi_ops_cq),
+	.read = ofi_cq_read,
+	.readfrom = ofi_cq_readfrom,
+	.readerr = ofi_cq_readerr,
+	.sread = fi_no_cq_sread,
+	.sreadfrom = fi_no_cq_sreadfrom,
+	.signal = fi_no_cq_signal,
+	.strerror = efa_cq_strerror
+};
+
+void efa_cq_progress(struct util_cq *cq)
+{
+	efa_cq_poll_ibv_cq(efa_env.efa_cq_read_size, cq);
+}
+
+static int efa_cq_close(fid_t fid)
+{
+	struct efa_cq *cq;
+	int ret;
+
+	cq = container_of(fid, struct efa_cq, util_cq.cq_fid.fid);
+
+	if (cq->ibv_cq.ibv_cq_ex) {
+		ret = -ibv_destroy_cq(ibv_cq_ex_to_cq(cq->ibv_cq.ibv_cq_ex));
+		if (ret) {
+			EFA_WARN(FI_LOG_CQ, "Unable to close ibv cq: %s\n",
+				fi_strerror(-ret));
+			return ret;
+		}
+		cq->ibv_cq.ibv_cq_ex = NULL;
+	}
+
+	ret = ofi_cq_cleanup(&cq->util_cq);
+	if (ret)
+		return ret;
+
+	free(cq);
+
+	return 0;
+}
+
+static struct fi_ops efa_cq_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = efa_cq_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+
+int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
+		struct fid_cq **cq_fid, void *context)
+{
+	struct efa_cq *cq;
+	struct efa_domain *efa_domain;
+	int err, retv;
+
+	if (attr->wait_obj != FI_WAIT_NONE)
+		return -FI_ENOSYS;
+
+	cq = calloc(1, sizeof(*cq));
+	if (!cq)
+		return -FI_ENOMEM;
+
+	err = ofi_cq_init(&efa_prov, domain_fid, attr, &cq->util_cq,
+			  &efa_cq_progress, context);
```

Comment:
You need a similar change for efa_cntr: https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/efa_cntr.c#L200-L202. Since you make dgram_ep_progress as a no op, `ofi_cntr_progress` will be a no-op too as it call ep progress in the same way as `ofi_cq_progress` does. You need a `efa_cntr_progress` that polls ibv_cq as well